refactor: rai_bench

jmatejcz opened the pull request:
## Purpose
Unify folder structure and naming of benchmarks 
Refactor tool_calling bench as there are conflicts between branches:
https://github.com/RobotecAI/rai/tree/mk/feat/spatial-reasoning-tasks
https://github.com/RobotecAI/rai/tree/jm/feat/tool-benchmark-custom-interfaces
which both changed the tool calling benchmark and were done in a hurry. That resulted in a lot of conflicts and not the best code.

Merge and unify taks that are already on development with spatial, navigation and custom interfaces tasks.
Improve and unify logging and saving results 

This PR is big in size, please follow below changes descriptions. Firstly focus on 1st and 2nd points. New frame  is essential in this PR, it dictates how validation is executed and a lot of other changes in this PR are adjusted to this frame.

Related PRs and branches will be also closed when this PR is merged.

> ⚠️ **Warning:** This PR does not provide  full benchmark, that is ready to test models. Only sample tasks were created. The action mocks/tests are also missing as I decided to do this in separate PRs.


## Proposed Changes
1. Restructured package:
 - all naming unified, now there the benchmarks: `manipulation_o3de` and `tool_calling_agent`. Every folder related to the benchmark will have exactly that name.
 - all code providing framework for creating benchmark in adequate folder
 - all code related to specific benchmark implementation in `examples/`  folder
 - experiment logs and results in `experiments/` folder
 - files/folders responsible for interfaces, tasks, benchmarks etc. named the same across benchmarks
 
              ├── rai_bench
              │ ├── examples/
              │ ├── experiments/
              │ ├── tool_calling_agent/
              │ ├── manipulation_o3de/
              

REST OF THE POINTS ARE FOR TOOL CALLING AGENT BENCHMARK:
2. New frame for tool_calling_agent benchmark:
- SubTask - smallest block, responsible for validating single tool call (ex. ListTopics)
- Validator - consists of subtasks. Based on the validator type, checks whether all subtasks were completed in a certain way
- Task - consists of validators. Every Validator can be treated as single step that is scored atomically. Visit `examples/tool_calling_agent/tasks.py` for more intuition. Every Task has always the same prompt and available tools, only the validation methods can be parametrized. On top of validator, you can pass extra_tool_calls param to allow model to correct itself.

![refactor_bench(3)](https://github.com/user-attachments/assets/87d39f93-6901-498b-aaef-b1bafe871977)

3. Migrated, Refactored and Unified tool_calling_agent benchmark tasks:
[tasks](https://github.com/RobotecAI/rai/tree/jm/refactor/rai_bench/src/rai_bench/rai_bench/tool_calling_agent/tasks)
- Tasks that were already on development - splited into basic and manipulation
- Custom interface tasks - from https://github.com/RobotecAI/rai/pull/487
- Spatial reasoning tasks - from https://github.com/RobotecAI/rai/pull/493
- Navigation tasks - from https://github.com/RobotecAI/rai/tree/mk/feat/tool-calling-bench-navigation-tasks

4.  Mocks 
- Mocks of tools (mostly imported from above mentioned branches)
- Refactored action mocks, but here more work is needed - will be continued in this issue -> https://github.com/RobotecAI/rai/issues/526

5. Models
[models](https://github.com/RobotecAI/rai/tree/jm/refactor/rai_bench/src/rai_bench/rai_bench/tool_calling_agent/messages)
- Pydantic models that reflect messages from ros2, which enables validation

6. Unit tests 
[tests](https://github.com/RobotecAI/rai/tree/jm/refactor/rai_bench/tests/rai_bench/tool_calling_agent)
for subtasks and validators

7. New GetInterfaceTool
old version didnt return types of fields
- imported from https://github.com/RobotecAI/rai/tree/jm/feat/tool-benchmark-custom-interfaces
- now returns one big string  as ros interface show cli command


8. Results and logs
- (Unchanged) all logs are logged to benhcmark.log files
- Result file is intended to be source of info for further processing
structure of results: 
![image](https://github.com/user-attachments/assets/1ef93cc3-f8a7-4716-8e11-10e4c7c19aba)
results now will have list of validators that will show what is expected by every validator
followed by passed list which holds bool for every validator
followed by score(which is redundant with passed, but its weird to not have score in results xd, if you have ideas here, please share)
followed by errors which is list of lists, where every validator has its own list of errors. 

9. Args passed when running tool calling agent benchmark
User can pass 2 args when running benchmark - model_name and vendor

10. Small docs [validation](https://github.com/RobotecAI/rai/blob/jm/refactor/rai_bench/docs/developer_guide/tool_calling_agent_benchmark_validation.md) as i wanted to paste image. This docs is small for now but i guess there will be docs for benchmarks in the future anyway.

11. Script to test multiple models, different benchamrks or couple repeats in one go https://github.com/RobotecAI/rai/blob/jm/refactor/rai_bench/src/rai_bench/rai_bench/examples/test_models.py

## Issues
[515](https://github.com/RobotecAI/rai/issues/515)
related PRs and branches:
https://github.com/RobotecAI/rai/pull/493
https://github.com/RobotecAI/rai/pull/487
[mk/feat/tool-calling-bench-navigation-tasks](https://github.com/RobotecAI/rai/tree/mk/feat/tool-calling-bench-navigation-tasks)


## Testing
Test single
```bash
python src/rai_bench/rai_bench/examples/tool_calling_agent/main.py --model-name llama3.2 --vendor ollama
```

tests:
```bash
pytest tests/rai_bench/
```

script running benchmarks:
```bash
python src/rai_bench/rai_bench/examples/test_models.py 
```

## next steps
- Actions mocks -> https://github.com/RobotecAI/rai/issues/526
- Full benchmark -> defining validators and tasks

MagdalenaKotynia committed:
feat: migrated spatial reasoning tasks

jmatejcz committed:
feat: GetROS2MessageInterfaceTool now return output as string with all subtypes and descriptions just like ros2 interface show command

jmatejcz committed:
refactor: restructure rai_bench

jmatejcz committed:
refacotr: separate tool_calling tasks

jmatejcz committed:
feat: add subtasks, validators
apply new frame to example tasks
changed results saving

jmatejcz commented:
@MagdalenaKotynia  @maciejmajek  please take a look, if you like this reconstruction of package structure and the new frame. If yes i will proceed with applying this refactor to other tasks and changes from 2 conflicted branches.

The refactor is not completed yet, i only applied the new frame to 2 task as an example, so don't pay attention to the not touched parts of code

jmatejcz committed:
refactor: renamed benchmar dirs

jmatejcz committed:
refactor: base class for basic tasks

jmatejcz committed:
feat: add more subtask to cover basic tests

jmatejcz committed:
feat add pydantic models for messages

Co-authored-by: Magdalena Kotynia <magdalena.kotynia@robotec.ai>

jmatejcz committed:
feat: add mock  tools for topics, services and actions

Co-authored-by: Magdalena Kotynia <magdalena.kotynia@robotec.ai>

jmatejcz committed:
feat: refactor maniupulaiton tasks to suit new frame

jmatejcz committed:
refactor: changed subtasks to more generic classes

jmatejcz commented:
How to log errors when extra calls passed? for example in 3 calls there are errors but in 4th agent done it correctly, should we      log the previous 3, even if validator passed eventually?

MagdalenaKotynia commented on src/rai_bench/rai_bench/tool_calling_agent/scores_tracing.py:
```diff
@@ -35,21 +35,21 @@ def get_callbacks() -> List[BaseCallbackHandler]:
                 trace_id=str(run_id),
                 name="tool calls result",
-                value=float(success),
+                value=float(score),
```
Score is already float, so you don't need to convert it to float.

jmatejcz replied:
fixed here: https://github.com/RobotecAI/rai/pull/517/commits/ef8ae2fddc526cbe35eb121fcfc9994d979ec48b

MagdalenaKotynia commented on src/rai_bench/rai_bench/tool_calling_agent/scores_tracing.py:
```diff
@@ -35,21 +35,21 @@ def get_callbacks() -> List[BaseCallbackHandler]:
                 run_id=run_id,
                 key="tool calls result",
-                score=float(success),
+                score=float(score),
```
same: https://github.com/RobotecAI/rai/pull/517/files#r2037418671

jmatejcz replied:
fixed here: https://github.com/RobotecAI/rai/pull/517/commits/ef8ae2fddc526cbe35eb121fcfc9994d979ec48b

jmatejcz committed:
refactor: changed errors and validation logging
improved csv prompts logging

MagdalenaKotynia commented on src/rai_bench/rai_bench/tool_calling_agent/interfaces.py:
```diff
@@ -0,0 +1,280 @@
+
+        done_properly = 0
+        for validator in self.validators:
+            if_success, remaining_tool_calls = validator.validate(
```
What about is_success?

jmatejcz replied:
you mean naming?

MagdalenaKotynia replied:
yes

MagdalenaKotynia commented on src/rai_bench/rai_bench/tool_calling_agent/interfaces.py:
```diff
@@ -0,0 +1,280 @@
+        for arg_name, arg_value in expected_args.items():
+            if arg_name in tool_call["args"]:
+                if tool_call["args"][arg_name] != arg_value:
+                    SubTaskValidationError(
```
```suggestion
                    raise SubTaskValidationError(
```

jmatejcz replied:
applied here: https://github.com/RobotecAI/rai/pull/517/commits/ef8ae2fddc526cbe35eb121fcfc9994d979ec48b

MagdalenaKotynia commented on src/rai_bench/rai_bench/tool_calling_agent/interfaces.py:
```diff
@@ -0,0 +1,280 @@
+                        f"Expected argument '{arg_name}' should have value '{arg_value}', but got '{tool_call['args'][arg_name]}'"
+                    )
+            else:
+                SubTaskValidationError(
```
```suggestion
                raise SubTaskValidationError(
```

jmatejcz replied:
applied here: https://github.com/RobotecAI/rai/pull/517/commits/ef8ae2fddc526cbe35eb121fcfc9994d979ec48b

MagdalenaKotynia commented on src/rai_bench/rai_bench/tool_calling_agent/interfaces.py:
```diff
@@ -0,0 +1,280 @@
+            if arg_name not in expected_args:
+                # If this argument is not required, check if it's an allowed optional argument
+                if not expected_optional_args or arg_name not in expected_optional_args:
+                    SubTaskValidationError(
```
```suggestion
                    raise SubTaskValidationError(
```

jmatejcz replied:
applied here: https://github.com/RobotecAI/rai/pull/517/commits/ef8ae2fddc526cbe35eb121fcfc9994d979ec48b

MagdalenaKotynia commented on src/rai_bench/rai_bench/tool_calling_agent/interfaces.py:
```diff
@@ -0,0 +1,280 @@
+                # If optional argument has expected value, check if the value is correct
+                elif expected_optional_args[arg_name]:
+                    if expected_optional_args[arg_name] != arg_value:
+                        SubTaskValidationError(
```
```suggestion
                        raise SubTaskValidationError(
```

jmatejcz replied:
applied here: https://github.com/RobotecAI/rai/pull/517/commits/ef8ae2fddc526cbe35eb121fcfc9994d979ec48b

MagdalenaKotynia commented on src/rai_bench/rai_bench/tool_calling_agent/validators.py:
```diff
@@ -0,0 +1,70 @@
+            except StopIteration:
+                return True, tool_calls[i:]
+
+        self.log_error(msg="Failed to validate")
```
I suggest logging less generic message with more info about what tool failed to validate

MagdalenaKotynia reviewed (commented):
The proposed new structure overall looks good to me. 
Please remember about handling strict/not strict validation. Currently, both implemented validators seem to be strict and not allow agent to self-correct. Please also remember to handle tool call args that may have allowed value in some range and situations when value does not matter.
FYI I didn't do code review, I just reviewed the structure, but I left some comments regarding the code when I noticed something by the way.

jmatejcz commented:
> The proposed new structure overall looks good to me. Please remember about handling strict/not strict validation. Currently, both implemented validators seem to be strict and not allow agent to self-correct. Please also remember to handle tool call args that may have allowed value in some range and situations when value does not matter. FYI I didn't do code review, I just reviewed the structure, but I left some comments regarding the code when I noticed something by the way.

i think you you misunderstood the validators-task relation. You can make validation strict or less strict by adjusting subtasks passed to validators and extra_tool_calls param, please look at `rai_bench/examples/tool_calling_agent/tasks.py`.

also look at the pic of framework:
https://github.com/RobotecAI/rai/issues/515

jmatejcz committed:
fix: raising errors SubTask
moved GetObjectPositionsTask to basics

jmatejcz committed:
fix: validator properly iterates when extra tools

jmatejcz committed:
fix: proper logging in manipulation tasks
typo in mocked tool

jmatejcz committed:
feat: add and intergarte custom interface tasks
add checking mesasge dicts subtasks

jmatejcz committed:
fix: import adjustment after rebase

jmatejcz committed:
feat: add action models

Co-authored-by: Magdalena Kotynia <magdalena.kotynia@robotec.ai>

jmatejcz committed:
refactor: changed available models in mock action tool

jmatejcz committed:
feat: add mock of GetDistanceToObjectsTool

Co-authored-by: Magdalena Kotynia <magdalena.kotynia@robotec.ai>

jmatejcz committed:
feat: add action toolkit mock

jmatejcz committed:
feat: check fields action subtask added

jmatejcz committed:
feat: migrated and refactored navigation tasks

Co-authored-by: Magdalena Kotynia <magdalena.kotynia@robotec.ai>

jmatejcz committed:
chore: deleted unused code

jmatejcz committed:
fix: handling recursion limit error

jmatejcz committed:
refactor: all message models now forbid extra params

jmatejcz committed:
feat: declared and moved sample tasks to one file

jmatejcz committed:
test: add unittests for subtasks
fixes to subtasks
import fix to old tests

jmatejcz committed:
test: add unittest for validators
fixes tot validators

jmatejcz committed:
feat: add args model_type and vendor

jmatejcz committed:
refactor: tests naming change

jmatejcz committed:
docs: added README and small validation docs

jmatejcz committed:
style: added  docstrings
deleted unused code

jmatejcz committed:
fix: fixes after rebase

jmatejcz committed:
chore: add licenses to tests

maciejmajek commented on src/rai_core/rai/tools/ros2/utils.py:
Answering your question, please rename the file to `interface_parser.py` and move it to the generic folder.

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/517/commits/678047ac5dfb5b154405bf3441aef2e0fd8858da

jmatejcz committed:
refactor: renamed interface parser file and moved it to generic folder

jmatejcz committed:
docs: moved docs to rai_bench

jmatejcz committed:
refactor: save to results errors from extra calls

jmatejcz committed:
fix: removed unnecessary configs froma action models

jmatejcz commented:
moved docs to rai_bench: https://github.com/RobotecAI/rai/pull/517/commits/bdc58e6fdc1adcbfce9385fc7f25609a5610353d

jmatejcz committed:
refactor: removed deafult empty dict from subtasks
adjusted error messages in validators

jmatejcz commented:
saving errors from extra tool calls: https://github.com/RobotecAI/rai/pull/517/commits/010fc5d466b40d1f1ee1e8862fb26257da31425a

jmatejcz commented:
removed default empty dicts from subtasks and improved error messages in validators:
https://github.com/RobotecAI/rai/pull/517/commits/710864f7d41165deb96c3d50a1026acddc5edfd5

jmatejcz committed:
feat: adjusted recursion limits
gathering tool calls even when recurrsion limit occurs

jmatejcz commented:
@maciejmajek  we speaked about setting recursion limit to number of tool calls required in task, but 1 step is just one node execution, not a tool call, so restricting it like this does not make much sense (Tool calls are in AIMessages).
![image](https://github.com/user-attachments/assets/56efc11a-157a-47f2-be8a-64b4b9cbb2f3)

I've set the recursion limit to 4*(required_tool_calls+extra_tool_calls), changed agent.invoke to stream which lets us collect messages from agent even when recursion limit occurs:  https://github.com/RobotecAI/rai/pull/517/commits/0cddef429700c424509a39d1ffa25ee1dbeec1c4

jmatejcz committed:
feat: added script for testing multiple models in one run
refactor running benchmarks to take args
added model initilization via model name

jmatejcz commented:
added python script that lets user run multiple models on different benchmarks in one go:
https://github.com/RobotecAI/rai/pull/517/commits/ce0979bc0d5d76ecd1b0eb2365b40f1ce9d6da44
Also modified model initialization to pass only model name and vendor.

Modified instructions in README : https://github.com/RobotecAI/rai/pull/517/commits/e6953e375466b7cc586f6f471ca4fa738da5cc13
and in PR description  on how to use it.

jmatejcz committed:
docs: adjusted README to new changes

jmatejcz committed:
fix: fixes  after rebase

maciejmajek commented on src/rai_core/rai/initialization/model_initialization.py:
```diff
@@ -157,6 +160,41 @@ def get_llm_model(
         raise ValueError(f"Unknown LLM vendor: {vendor}")
 
 
+def get_llm_model_direct(
+    model_name: str,
+    vendor: str,
+    config_path: Optional[str] = None,
+    **kwargs: Any,
+) -> ChatOpenAI | ChatBedrock | ChatOllama:
+    config = load_config(config_path)
+    model_config = getattr(config, vendor)
+
+    logger.info(f"Initializing Model: {model_name}, Vendor: {vendor}")
+    if vendor == "openai":
+        from langchain_openai import ChatOpenAI
+
+        model_config = cast(OpenAIConfig, model_config)
+
+        return ChatOpenAI(model=model_name, base_url=model_config.base_url, **kwargs)
+    elif vendor == "aws":
+        from langchain_aws import ChatBedrock
+
+        model_config = cast(AWSConfig, model_config)
+
+        return ChatBedrock(
+            model_id=model_name,
+            region_name=model_config.region_name,
+            **kwargs,
+        )
+    elif vendor == "ollama":
+        from langchain_ollama import ChatOllama
+
+        model_config = cast(OllamaConfig, model_config)
+        return ChatOllama(model=model_name, base_url=model_config.base_url, **kwargs)
+    else:
+        raise ValueError(f"Unknown LLM vendor: {vendor}")
+
+
```
This could be a factory, apart from that lgtm
